File: streamcipher2.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#define WORDSIZE unsigned char
#define WORDSIZE_BITS 8
#define ROTATIONS 5
#define PRINT_STRING "%i: %u\n"
#define STATESIZE 32

#ifndef memcpy_s
void memcpy_s(WORDSIZE* s1, WORDSIZE* s2, size_t n)
{
    unsigned int index;
    for (index = 0; index < n; index++)
    {
        s1[index] = s2[index];
    }
}
#endif

void print_data(WORDSIZE* data)
{
    int index;
    printf("\n");
    for (index = 0; index < 16; index++)
    {
        printf(PRINT_STRING, index, data[index]);
    }
} 

WORDSIZE rotate_left(WORDSIZE word, int amount)
{    
    return ((word << amount) | (word >> (WORDSIZE_BITS - amount)));
}

WORDSIZE rotate_right(WORDSIZE word, int amount)
{    
    return ((word >> amount) | (word << (WORDSIZE_BITS - amount)));
}

void shuffle_bytes(WORDSIZE* _state)
{
    WORDSIZE temp[16];
        
    temp[7]  = _state[0];
    temp[12] = _state[1];
    temp[14] = _state[2];
    temp[9]  = _state[3];
    temp[2]  = _state[4];
    temp[1]  = _state[5];
    temp[5]  = _state[6];
    temp[15] = _state[7];
    temp[11] = _state[8];
    temp[6]  = _state[9];
    temp[13] = _state[10];
    temp[0]  = _state[11];
    temp[4]  = _state[12];
    temp[8]  = _state[13];
    temp[10] = _state[14];
    temp[3]  = _state[15];
    
    memcpy_s(_state, temp, 16);           
}    
    
int prp(WORDSIZE* data, WORDSIZE key, int transpose, unsigned long datasize)
{
    WORDSIZE left, right;    
    int index;
    if (transpose)
    {
        shuffle_bytes(data);
        shuffle_bytes(data+16);
    }
    
    for (index = datasize - 2; index != -1; index--)
    {                                    
        
        right = data[index + 1];
        left = data[index];
        
        key ^= right;                       
        right = rotate_left((right + key + index), ROTATIONS);            
        key ^= right;                       
        
        key ^= left;        
        left = (left + (right >> (WORDSIZE_BITS / 2)));        
        left ^= rotate_left(right, (index % WORDSIZE_BITS) ^ ROTATIONS);                
        key ^= left;
                
        data[index + 1] = right;        
        data[index] = left;
    }
    
    key ^= data[0];
    data[0] = data[0] + key;
    key ^= data[0];        
    return key;
}
        
void prf(WORDSIZE* data, WORDSIZE key, WORDSIZE data_size)
{
    WORDSIZE index, byte;
    for (index = 0; index < data_size; index++)
    {    
        byte = rotate_left((data[index] + key + index), ROTATIONS);  
        key ^= byte;
        data[index] = byte;           
    }
}
    
WORDSIZE xor_with_key(WORDSIZE* data, WORDSIZE* key, unsigned int datasize)
{
    WORDSIZE index, new_key = 0;
    for (index = 0; index < datasize; index++)
    {           
        data[index] ^= key[index];   
        new_key ^= data[index];
    }        
    return new_key;
}

void stream_cipher(WORDSIZE* key_material, WORDSIZE* seed, WORDSIZE* _key, unsigned long blocks)
{
    WORDSIZE state[32];   
    WORDSIZE state_xor = 0, key_material_xor = 0;
    
    unsigned long index;
    
    for (index = 0; index < 16; index++)
    {                
        state[index] = seed[index];        
        state[index + 16] = _key[index];
        state_xor ^= state[index] ^ state[index + 16];
    }                         
        
    for (index = 0; index < blocks; index++)
    {                             
        state_xor = prp(state, state_xor, 1, STATESIZE);
        key_material_xor ^= state_xor;        
        memcpy_s(key_material + (index * 16), state, 16);        
    }
    
    prf(key_material, key_material_xor, blocks * 16);           
}
  
void encrypt(WORDSIZE* data, WORDSIZE* key, WORDSIZE* seed, unsigned long data_size)
{    
    unsigned long blocks, extra;  
    WORDSIZE state_xor;
    blocks = data_size / 16;
    extra = data_size % 16;
    if (extra)
    {
        blocks += 1;
    }

    unsigned long byte_count = blocks * 16 * sizeof(WORDSIZE);
    WORDSIZE* key_material = malloc(byte_count);
    
    stream_cipher(key_material, seed, key, blocks);   
    
    state_xor = xor_with_key(data, key_material, byte_count);    
    prp(data, state_xor, 0, byte_count);
    xor_with_key(data, key_material, byte_count);
    
    free(key_material);    
}

void invert_shuffle_bytes(WORDSIZE* state)
{
    WORDSIZE temp[16];
    
    temp[11] = state[0];
    temp[5]  = state[1];
    temp[4]  = state[2];
    temp[15] = state[3];
    temp[12] = state[4];
    temp[6]  = state[5];
    temp[9]  = state[6];
    temp[0]  = state[7];
    temp[13] = state[8];
    temp[3]  = state[9];
    temp[14] = state[10];
    temp[8]  = state[11];
    temp[1]  = state[12];
    temp[10] = state[13];
    temp[2]  = state[14];
    temp[7]  = state[15];
    
    memcpy_s(state, temp, 16);    
}

WORDSIZE invert_prp(WORDSIZE* data, WORDSIZE key, int transpose, unsigned long data_size)
{
    WORDSIZE left, right;    
    int index;
    
    key ^= data[0];
    data[0] = data[0] - key;
    key ^= data[0];

    for (index = 0; index < data_size - 1; index++)
    {                                    
        
        right = data[index + 1];
        left = data[index];
        
        key ^= left;
        left ^= rotate_left(right, (index % WORDSIZE_BITS) ^ ROTATIONS);
        left -= right >> (WORDSIZE_BITS / 2);
        key ^= left;
        
        key ^= right;                       
        right = rotate_right(right, ROTATIONS) - key - index;            
        key ^= right;                               
                
        data[index + 1] = right;        
        data[index] = left;
    }

    if (transpose)
    {
        invert_shuffle_bytes(data);
        invert_shuffle_bytes(data+16);
    }
    
    return key;
}

void decrypt(WORDSIZE* data, WORDSIZE* key, WORDSIZE* seed, unsigned long data_size)
{    
    unsigned long blocks, extra;  
    WORDSIZE state_xor;
    blocks = data_size / 16;
    extra = data_size % 16;
    if (extra)
    {
        blocks += 1;
    }

    unsigned long byte_count = blocks * 16 * sizeof(WORDSIZE);
    WORDSIZE* key_material = malloc(byte_count);
    
    stream_cipher(key_material, seed, key, blocks);   
    
    state_xor = xor_with_key(data, key_material, byte_count);    
    invert_prp(data, state_xor, 0, byte_count);
    xor_with_key(data, key_material, byte_count);
    
    free(key_material);    
}
    
void test_encrypt_decrypt()
{    
    WORDSIZE data[16], key[16], plaintext[16], null_string[16], seed[16];    
    
    memset(null_string, 0, 16);
    memcpy_s(data, null_string, 16);
   // memcpy_s(data+16, null_string, 16);
    //data[15] = 1;
    memcpy_s(plaintext, data, 16);
    memcpy_s(key, null_string, 16); 
    memcpy_s(seed, null_string, 16);
   
    //WORDSIZE xor = 0;
    //xor = prp(data, 0, 1, 32);
    //prf(data, xor, 32);
    //printf("Data:\n%s\n", data);
    //print_data(data);
    
    encrypt(data, key, seed, 16);
    
    printf("Ciphertext:\n %s\n", data);    
    print_data(data);
        
    decrypt(data, key, seed, 16);
    printf("Decrypted:\n");
    print_data(data);    
}

int main()
{
    test_encrypt_decrypt();            
    return 0;
}

File: designs/streamcipher/streamcipher2.py
from crypto.utilities import xor_sum, rotate_left, xor_subroutine, modular_subtraction, rotate_right, words_to_bytes

def shuffle_bytes(_state, section, temp=list(range(16))):          
    temp[7] = _state[0] 
    temp[12] = _state[1]
    temp[14] = _state[2]
    temp[9] = _state[3]
    temp[2] = _state[4]
    temp[1] = _state[5]
    temp[5] = _state[6]
    temp[15] = _state[7]
    temp[11] = _state[8]
    temp[6] = _state[9]
    temp[13] = _state[10]
    temp[0] = _state[11]
    temp[4] = _state[12]
    temp[8] = _state[13]
    temp[10] = _state[14]
    temp[3] = _state[15]
            
    _state[section] = temp[:]
           
def prp(data, key, mask=255, rotation_amount=5, bit_width=8, key_slice=slice(16, 32), data_slice=slice(0, 16), transpose=True):     
    if transpose:
        if key_slice is not None:
            shuffle_bytes(data, key_slice)
        shuffle_bytes(data, data_slice)       
    
    for index in reversed(range(len(data) - 1)):       
        left, right = data[index], data[index + 1]
        
        key ^= right                 
        right = rotate_left((right + key + index) & mask, rotation_amount, bit_width)                
        key ^= right
        
        key ^= left        
        left = (left + (right >> (bit_width / 2))) & mask                
        left ^= rotate_left(right, (index % bit_width) ^ rotation_amount)                    
        key ^= left
        
        data[index], data[index + 1] = left, right                 

    right = data[0]
    key ^= right
    right = (right + key) & mask
    key ^= right
    
    left = data[1]
    key ^= left        
    left = (left + (right >> (bit_width / 2))) & mask                
    left ^= rotate_left(right, (index % bit_width) ^ rotation_amount)                    
    key ^= left
    data[1] = left
    return key
        
def invert_prp(data, key, mask=255, rotation_amount=5, bit_width=8, transpose=False):           
    key ^= data[0]
    data[0] = (256 + (data[0] - key)) & mask
    key ^= data[0]
    
    for index in range(len(data) - 1):
        
        right = data[index + 1]
        left = data[index]
        
        key ^= left    
        left ^= rotate_left(right, (index % bit_width) ^ rotation_amount)         
        left = (256 + (left - (right >> bit_width / 2))) & mask
        key ^= left;
        
        key ^= right;        
        right = (256 + (rotate_right(right, rotation_amount, bit_width) - key - index)) & mask 
        key ^= right;                               
                
        data[index + 1] = right;        
        data[index] = left;
    
    return key;      
            
def prf(data, key, mask=255, rotations=5, bit_width=8):      
    for index in range(len(data)):         
        new_byte = rotate_left((data[index] + key + index) & mask, rotations, bit_width)
        key ^= new_byte        
        data[index] = new_byte             
    
def xor_with_key(data, key):
    data_xor = 0
    for index, byte in enumerate(key):
        data[index] ^= byte
        data_xor ^= data[index]
    return data_xor
    
def stream_cipher(data, seed, key, size=(8, 255, 5), mode="encrypt"):     
    if mode == "encrypt":
        data_prp_function = prp
    else:
        data_prp_function = invert_prp
    key = list(key)
    seed = list(seed)
    state = seed + key
    key_material = list()       
    
    bit_width, mask, rotation_amount = size          
    block_count, extra = divmod(len(data), 16) 
    prf_state_xor = 0    
    
    state_xor = xor_sum(state)
    
    for block in range(block_count + 1 if extra else block_count):       
        state_xor = prp(state, state_xor, mask, rotation_amount, bit_width)
        prf_state_xor ^= state_xor                
        key_material.extend(state[0:16])    
    
    prf(key_material, prf_state_xor, mask, rotation_amount, bit_width)        
    
    data_xor = xor_with_key(data, key_material)        
    data_xor = data_prp_function(data, data_xor, mask, rotation_amount, bit_width, transpose=False)    
    xor_subroutine(data, key_material)
    
def test_prp_prf():
    data = bytearray(32)
    xor = prp(data, 0)    
    print data
    print [byte for byte in data]
    invert_prp(data, xor)
    print data
    print [byte for byte in data]
    
    #from differential import find_best_differential
    #from linear import calculate_linearity
    #for index in range(16):
    #    sbox = bytearray()        
    #    for byte in range(256):
    #        data = bytearray(32)
    #        data[15] = byte
    #        data[-2] = 1
    #        prp(data, byte)
    #        prf(data, xor_sum(data))
    #        sbox.append(data[index])     
    #    import os
    #    sbox = bytearray(os.urandom(256))
    #    from scratch import aes_s_box as sbox
    #    print "Calculating best differential/linearity for byte at index: {}...".format(index)        
    #    print find_best_differential(sbox)
    #    print calculate_linearity(sbox)
    
def test_streamcipher2():
    data = list(bytearray(16))
    key = bytearray(16)
    seed = bytearray(16)    
    size = (8, 255, 5)#(64, 0xFFFFFFFFFFFFFFFF, 40)
    data[1] = 1
    stream_cipher(data, key, seed, size)
    print words_to_bytes(data, size[0] / 8)
   # print [byte for byte in data]
    
    
    data2 = list(bytearray(16))
    data2[0] = 1
    stream_cipher(data2, key, seed, size, mode="encrypt")
    print words_to_bytes(data2, size[0] / 8)
   # print [byte for byte in data2]
        
if __name__ == "__main__":
    #test_prp_prf()
    test_streamcipher2()
    